Gt plt summary customization

ScottFB101 opened the pull request:
# Summary

Thank you for contributing to **gt-extras**! To make this process easier for everyone, please explain the context and purpose of your contribution. Also, list the changes made to the existing code or documentation.

This PR addresses the 4 out of the 5 enhancements proposed in Issue #146. The one enhancement not addressed was: "Display times instead of just dates". I felt that was deserving of another separate PR. 

The changes made to the existing code are rather straightforward, and don't need much background besides knowing they address the 4 enhancements previously referenced. This is my opinion of course, and I'm happy to elaborate more on the code here if there's a want for it.

As for the context and purpose of my contribution, I use so many posit-backed packages and tools, that it only feels right to get involved in contributing to the ecosystem. 

# Related GitHub Issues and PRs

- Ref: #146

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct/).
- [x] I have followed the [Style Guide for Python Code](https://peps.python.org/pep-0008/) as best as possible for the submitted code.
- [x] I have added **pytest** unit tests for any new functionality.

ScottFB101 committed:
allow hiding of Mean, Median, and SD columns

ScottFB101 committed:
Add mode calculation to summary table

Introduces a 'Mode' column to the summary table in gt_plt_summary, with logic to handle cases with no singular mode or multiple modes. Updates function arguments to allow hiding descriptive stats and mode, and adjusts .gitignore to exclude VSCode launch configuration.

ScottFB101 committed:
Refactor summary stats column handling

Mode column is now conditionally dropped based on the add_mode flag, and descriptive stats columns are managed more flexibly. This improves clarity and control over which summary statistics are displayed.

ScottFB101 committed:
Descriptions of new boolean variables

ScottFB101 committed:
Add interactivity option to summary plots

Introduces an 'interactivity' parameter to summary plot functions, allowing toggling of interactive features such as hover tooltips and CSS. All relevant plotting and SVG generation functions now accept and respect this parameter, enabling more flexible plot rendering.

NOTE: The code that creates the tooltips still runs, it just isn't append to the SVG Element.

ScottFB101 committed:
Add customizable color mapping to gt_plt_summary

Introduces a color_mapping parameter to gt_plt_summary, allowing users to override the default color scheme. A helper function change_color_mapping updates the global COLOR_MAPPING when user overrides are provided.

ScottFB101 committed:
Refactor summary stats and enable plot interactivity

Refactored summary DataFrame creation to conditionally include descriptive statistics and mode based on parameters. Enabled interactivity by default in category bar and histogram SVG plots. Commented out redundant DataFrame column drops in gt_plt_summary.

ScottFB101 committed:
Enable interactivity by default in summary plots. Added interactivity to boolean and numeric plots.

Changed the default value of the 'interactivity' parameter to True in all summary plot functions to ensure interactive plots are generated by default.

Also

ScottFB101 committed:
Add tests for change_color_mapping function

Added unit tests to verify that change_color_mapping correctly updates COLOR_MAPPING, handles None and empty dict inputs, and preserves existing keys. This improves test coverage for color mapping customization.

ScottFB101 committed:
Fix column existence checks for summary formatting

Adds checks to ensure that missing value handling and number formatting are only applied to columns present in the DataFrame. This prevents errors in Polars when columns like 'Mean', 'Median', 'SD', or 'Mode' are absent.

ScottFB101 committed:
Remove .vscode/launch.json from .gitignore

The .vscode/launch.json entry was deleted from .gitignore

ScottFB101 committed:
Refactor color mapping override in gt_plt_summary

Removed the change_color_mapping function and integrated color mapping overrides directly into gt_plt_summary via the new_color_mapping parameter. Updated related docstrings and cleaned up unused code and tests for change_color_mapping.

ScottFB101 committed:
Document interactivity parameter in gt_plt_summary

Added documentation for the 'interactivity' parameter in the gt_plt_summary function, explaining its purpose for toggling interactive features in Plot Overview column graphs.

juleswg23 commented on gt_extras/summary.py:
```diff
@@ -50,6 +57,33 @@ def gt_plt_summary(df: IntoDataFrame, title: str | None = None) -> GT:
+                "string": "purple",
+                "numeric": "pink",
+            }
+
```
Really appreciate the new docstring!

This is maybe a little nit-picky, but I think we can add an example with `new_color_mapping` to the main docstring examples section, and leave it out here. Instead you could describe what the dictionary is. Something like: `"A dictionary where the keys are ... and the values are ..."`

One way to generate the docs is by going into the root directory of the package and running `make docs-build` to build a local version of the documentation. Once you've done that once, you should be able to navigate into the docs directory (`cd docs`) and execute `quartodoc build; quarto preview`, to quickly see a live version of the API reference for this function.

juleswg23 commented on gt_extras/summary.py:
```diff
@@ -50,6 +57,33 @@ def gt_plt_summary(df: IntoDataFrame, title: str | None = None) -> GT:
+        Boolean that allows the hiding of the Mean, Median, and SD columns. Defaults to 'False'.
+
+    add_mode
+        Boolean that allows the addition of a Mode column. Defaults to 'False'.
```
We can take advantage of quarto to freshen up the docstrings. Using backticks (```` `False` ```` instead of ```` 'False' ```` or ```` `"string"` ```` instead of ```` "String" ````) will render the text as code.

ScottFB101 committed:
Rename hide_desc_stats to show_desc_stats in summary

Replaces the 'hide_desc_stats' parameter with 'show_desc_stats' in gt_plt_summary and related functions for clarity. Updates logic and docstrings to reflect the new parameter name and default value.

ScottFB101 committed:
new_color_mapping example to main docstring

Improves the docstring for gt_plt_summary by clarifying the new_color_mapping parameter as a dictionary mapping data types to hex color codes, and adds a new example demonstrating custom color mapping with ocean swell data. Also corrects default value descriptions for show_desc_stats and add_mode.

ScottFB101 committed:
Show category tooltips and bars only with interactivity

Moved the creation and insertion of visual bars and tooltips inside the interactivity check in _make_categories_bar_svg. This ensures that these SVG elements are only added when interactivity is enabled, preventing unnecessary elements in static mode.

ScottFB101 committed:
Fix tooltip rendering logic in SVG bar and histogram

Adjusts the placement and conditional rendering of tooltips and visual bars in the _make_categories_bar_svg and _make_histogram_svg functions. Tooltips are now only created and appended when interactivity is enabled, preventing unnecessary elements in non-interactive SVGs.

ScottFB101 committed:
Struggling with the test here. Could use some advice

ScottFB101 commented:
Of course! Happy to help, it's been good practice for me.

Alright I made all the changes, but will admit I'm struggling a bit figuring out the snapshot test (not something I've dealt with before), so I can't get 2 of the 500ish lines of `summary.py` to get coverage using pytest.

juleswg23 commented:
> I can't get 2 of the 500ish lines of `summary.py` to get coverage using pytest.

This is getting close! The snapshot test is great, it's not necessarily supposed to hit every branch, just to check a standard example's output. If you run `make test-update` and commit the `test_summary.ambr` file in the `tests/__snapshots/` directory, those snapshot tests should pass with all future as you have them now in all future `make test` executions.

To get coverage on the un-hit lines (for example, I am seeing the below line isn't covered by the snapshot) you can write tests similar to the rest of the ones in `test_summary.py`. So for example, to test the below branch:

```python
# Limiting the number of modes displayed to two at maximum
elif len(mode_val) > 2:
    mode_val = "Greater than 2 Modes"
# Converting to string, then listing together
else:
    mode_val = ", ".join(str(i) for i in sorted(mode_val.to_list()))
```

We can use tests like:

``` python

@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_two_modes(DataFrame):
    df = DataFrame({"numeric": [1, 1, 2, 2, 3]})

    result = gt_plt_summary(df, add_mode=True)
    html = result.as_raw_html()

    assert '<td class="gt_row gt_left">1, 2</td>' in html

@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_greater_than_two_modes(DataFrame):
    df = DataFrame({"numeric": [1, 1, 2, 2, 3, 4, 4]})

    result = gt_plt_summary(df, add_mode=True)
    html = result.as_raw_html()

    assert '<td class="gt_row gt_left">Greater than 2 Modes</td>' in html
```

I know it's not the flashiest to do, but it would be nice to have these kinds of specific tests for each individual new parameter, and then optionally any edge cases you might want to include. (For example, establishing behavior that is expected when `add_mode=True` and `show_desc_stats=False`.)

ScottFB101 committed:
remove unnecessary defaults

ScottFB101 committed:
sorted numbers in mode column

ScottFB101 committed:
Add tests for multiple modes in gt_plt_summary

Introduces tests to verify gt_plt_summary correctly handles cases with two modes and more than two modes in both pandas and polars DataFrames.

ScottFB101 commented:
> > I can't get 2 of the 500ish lines of `summary.py` to get coverage using pytest.
> 
> ```python
> # Limiting the number of modes displayed to two at maximum
> elif len(mode_val) > 2:
>     mode_val = "Greater than 2 Modes"
> # Converting to string, then listing together
> else:
>     mode_val = ", ".join(str(i) for i in sorted(mode_val.to_list()))
> ```
> 
> We can use tests like:
> 
> ```python
> @pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
> def test_gt_plt_summary_two_modes(DataFrame):
>     df = DataFrame({"numeric": [1, 1, 2, 2, 3]})
> 
>     result = gt_plt_summary(df, add_mode=True)
>     html = result.as_raw_html()
> 
>     assert '<td class="gt_row gt_left">1, 2</td>' in html
> 
> @pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
> def test_gt_plt_summary_greater_than_two_modes(DataFrame):
>     df = DataFrame({"numeric": [1, 1, 2, 2, 3, 4, 4]})
> 
>     result = gt_plt_summary(df, add_mode=True)
>     html = result.as_raw_html()
> 
>     assert '<td class="gt_row gt_left">Greater than 2 Modes</td>' in html
> ```

Unfortunately Pandas handles the Mode column as an object. I guess there's no native string column type? Polars can handle string columns, so the alignment in the data cell varies depending on dataframe type. For example, Pandas results in "gt_right", Polars in "gt_left".

juleswg23 commented on gt_extras/summary.py:
```diff
@@ -113,13 +134,54 @@ def gt_plt_summary(df: IntoDataFrame, title: str | None = None) -> GT:
-    summary_df = _create_summary_df(df)
+    if new_color_mapping:
+        global COLOR_MAPPING
+        COLOR_MAPPING.update(new_color_mapping)
```
I don't think it solves the root of the problem.

If I run these three in succession, the third one is colored red, rather than orange.

```python
import pandas as pd
from gt_extras import gt_plt_summary

df = pd.DataFrame({"numeric": [1, 1, 2, 2, 3]})
gt_plt_summary(df)
```

``` python
df = pd.DataFrame({"numeric": [1, 1, 2, 2, 3]})
gt_plt_summary(df, new_color_mapping={"numeric":"red"})
```

``` python
df = pd.DataFrame({"numeric": [1, 1, 2, 2, 3]})
gt_plt_summary(df)
```

<img width="400" alt="image" src="https://github.com/user-attachments/assets/eb4779fc-30f2-4390-9b17-9c6dad258ff4" />

ScottFB101 replied:
Must of commented in the wrong spot originally, so there may be a similar response in this PR thread.

I could make a Pytest fixture that resets the color mapping to the preset from the `summary.py` script prior to each test.

juleswg23 commented:
> Unfortunately, Pandas handles the Mode column as an object. I guess there's no native string column type? Polars can handle string columns, so the alignment in the data cell varies depending on dataframe type. For example, Pandas results in "gt_right", Polars in "gt_left".
> 
> For the time being, I strictly checked that the correct values were in the data cell, but didn't check alignment.

Yeah, I'm seeing this too. I'd guess there's some workaround within the dataframe, but we could supersede it and set the following:

``` python
gt = gt.cols_align(align="right", columns="Mode")
```

Just note that's not going to work if the mode column isn't present.

I also don't think it's that bad to leave it as is. In that case, the user can edit the table if they see fit.

ScottFB101 committed:
Update summary tests to check for exact HTML output

Changed assertions in test_gt_plt_summary_two_modes and test_gt_plt_summary_greater_than_two_modes to verify the exact HTML table cell output instead of just substring matches. This ensures the tests are more robust and accurately reflect the rendered HTML structure.

ScottFB101 committed:
Align 'Mode' column to right in summary table

Adds right alignment to the 'Mode' column in the summary table when add_mode is enabled

juleswg23 committed:
replace snap test with parameterized test

juleswg23 committed:
propogate color mapping

juleswg23 committed:
run make-update

juleswg23 committed:
Merge branch 'main' into gt_plt_summary_customization
